simson/common/data_transformations.py

```diff
@@ -0,0 +1,123 @@
+from copy import copy
+import numpy as np
+from scipy.optimize import least_squares
+
+from sodym import (
+    StockArray, DynamicStockModel, FlowDrivenStock,
+    DimensionSet, NamedDimArray, Process, Parameter
+)
+
+
+def get_new_array(dims: DimensionSet, **kwargs) -> NamedDimArray:
+    dims = copy(dims).get_subset(kwargs["dim_letters"]) if "dim_letters" in kwargs else dims
+    return NamedDimArray(dims=dims, **kwargs)
+
+
+def get_subset_transformer(dims: DimensionSet, dim_letters: tuple):
+    """Get a Parameter/NamedDimArray which transforms between two dimensions, one of which is a subset of the
+    other."""
+    assert len(dim_letters) == 2, "Only two dimensions are allowed"
+    dims = copy(dims).get_subset(dim_letters)
+    assert set(dims[0].items).issubset(set(dims[1].items)) or set(dims[1].items).issubset(
+        set(dims[0].items)
+    ), f"Dimensions '{dims[0].name}' and '{dims[1].name}' are not subset and superset or vice versa."
+    out = NamedDimArray(name=f"transform_{dims[0].letter}_<->_{dims[1].letter}", dims=dims)
+    # set all values to 1 if first axis item equals second axis item
+    for i, item in enumerate(dims[0].items):
+        if item in dims[1].items:
+            out.values[i, dims[1].index(item)] = 1
+    return out    
+
+
+def extrapolate_stock(
+        historic_stocks: StockArray, dims: DimensionSet,
+        parameters: dict[str, Parameter], curve_strategy: str
+        ):
+    """Performs the per-capita transformation and the extrapolation."""
+
+    # transform to per capita
+    pop = parameters['population']
+    transform_t_thist  = get_subset_transformer(dims=dims, dim_letters=('t', 'h'))
+    historic_pop       = get_new_array(dims=dims, dim_letters=('h','r'))
+    historic_gdppc     = get_new_array(dims=dims, dim_letters=('h','r'))
+    historic_stocks_pc = get_new_array(dims=dims, dim_letters=('h','r','g'))
+    stocks_pc          = get_new_array(dims=dims, dim_letters=('t','r','g'))
+    stocks             = get_new_array(dims=dims, dim_letters=('t','r','g'))
+
+    historic_pop[...] = pop * transform_t_thist
+    historic_gdppc[...] = parameters['gdppc'] * transform_t_thist
+    historic_stocks_pc[...] = historic_stocks / historic_pop
+
+    if curve_strategy == "GDP_regression":
+        gdp_regression(historic_stocks_pc.values, parameters['gdppc'].values, stocks_pc.values)
+    else:
+        raise RuntimeError(f"Prediction strategy {curve_strategy} is not defined. "
+                           f"It needs to be 'GDP_regression'.")
+
+    # transform back to total stocks
+    stocks[...] = stocks_pc * pop
+
+    #visualize_stock(self, self.parameters['gdppc'], historic_gdppc, stocks, historic_stocks, stocks_pc, historic_stocks_pc)
+    return StockArray(**dict(stocks))
+
+
+def gdp_regression(historic_stocks_pc, gdppc, prediction_out):
```

Comment:
This should be in a separate file. There will be more functions with the same task but different methods. 

And so should extrapolate_stock.

Reply:
Okay, I don't think this should prevent the pr being merged, as we can separate the functions later when we add the extra cases (and if we have multiple options for the same task it might make sense to have a class structure where we can specify input and output). Did you manage to go through the rest of the pr, are you otherwise happy?